sklearn/cluster/k_means_.py

```diff
@@ -1282,15 +1290,20 @@ def fit(self, X, y=None):
             Coordinates of the data points to cluster
         """
         random_state = check_random_state(self.random_state)
-        X = check_array(X, accept_sparse="csr", order='C', dtype=np.float64)
+        # to handle sparse data which only works as float64 at the moment
+        if sp.issparse(X):
+            X = check_array(X, accept_sparse="csr", order='C',
+                            dtype=np.float64)
+        else:
+            X = check_array(X, accept_sparse="csr", order='C')
```

Comment:
`accept_sparse` is useless on this side of the if branch. Instead you should put:  `dtype=[np.float64, np.float32]` to make it possible to convert integer to `np.float64` by default while letting `np.float32` pass through unchanged.